File: src/apollo_t.mjs
import { BoolT, FloatT, IntT, StringT } from "./apollo/util.mjs";
import { Error, Ok } from "./gleam.mjs";

/** @type {Record<string, (v: unknown) => boolean>} */
const _validator_map = {
  Int: Number.isInteger,
  String: (v) => typeof v === "string",
  Float: (v) => typeof v === "number" && !Number.isInteger(v),
  Bool: (v) => typeof v === "boolean",
};

export class Value {
  /**
   * @param {"Int" | "String" | "Float" | "Bool"} type
   */
  constructor(type) {
    this.type = type;
    /** @type {Array<(v: any) => import('./gleam').Result<undefined, string>>} */
    this.checks = [];
  }

  /**
   * @param {any} obj
   * @returns {import('./gleam').Result<undefined, string>}
   */
  validate(obj) {
    // @ts-ignore
    if (!_validator_map[this.type](obj)) {
      return new Error(`Type of value \`${obj}\` is not \`${this.type}\``);
    }
    let res;
    for (const check of this.checks) {
      res = check(obj);
      if (!res.isOk()) return res;
    }
    // @ts-ignore
    return new Ok();
  }
}

export function new_value(type) {
  return new Value(type);
}

/**
 * @param {Value} v
 * @param {(v: any) => import('./gleam').Result<undefined, string>} validator
 */
export function custom(v, validator) {
  v.checks.push((o) => {
    let value;
    if (v.type === "Bool") value = new BoolT(o);
    else if (v.type === "Float") value = new FloatT(o);
    else if (v.type === "Int") value = new IntT(o);
    else if (v.type === "String") value = new StringT(o);
    return validator(value);
  });

  return v;
}

File: src/apollo_t.erl
-module(apollo_t).
-export([new_value/1, custom/2]).
-export_type([value/0]).

-type value() :: {value, atom(), list()}.

-spec new_value(Str :: binary()) -> value().
new_value(Str) ->
    {value, Str, []}.

-spec custom(V :: value(), Validator :: fun((any()) -> {ok, nil} | {error, binary()})) -> value().
custom({Value, Type, List}, Validator) ->
    {Value, Type,
        lists:append([
            List, [validation_factory(Type, Validator)]
        ])}.

validation_factory(Type, Validator) ->
    fun(O) ->
        case Type of
            <<"Bool">> -> Validator({bool_t, O});
            <<"Float">> -> Validator({float_t, O});
            <<"Int">> -> Validator({int_t, O});
            <<"String">> -> Validator({string_t, O});
            _ -> {error, <<"Something went wrong">>}
        end
    end.
